File: test/language/structure_types/cpp/FieldConstructorClashingTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

#include "structure_types/field_constructor_clashing/FieldConstructorClashing.h"

namespace structure_types
{
namespace field_constructor_clashing
{

class FieldConstructorClashingTest : public ::testing::Test
{
protected:
    void fillFieldConstructorClashing(FieldConstructorClashing& fieldConstructorClashing)
    {
        auto& compoundReadArray = fieldConstructorClashing.getCompoundReadArray();
        compoundReadArray.emplace_back(CompoundRead{Field{FIELD1}, Field{FIELD2}});

        auto& compoundPackingArray = fieldConstructorClashing.getCompoundPackingArray();
        compoundPackingArray.emplace_back(CompoundPackingRead{Field{FIELD1}, Field{FIELD2}, Field{FIELD3}});
    }

    static const uint32_t FIELD1;
    static const uint32_t FIELD2;
    static const uint32_t FIELD3;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const uint32_t FieldConstructorClashingTest::FIELD1 = 1;
const uint32_t FieldConstructorClashingTest::FIELD2 = 9;
const uint32_t FieldConstructorClashingTest::FIELD3 = 5;

TEST_F(FieldConstructorClashingTest, writeRead)
{
    FieldConstructorClashing fieldConstructorClashing;
    fillFieldConstructorClashing(fieldConstructorClashing);

    zserio::BitStreamWriter writer(bitBuffer);
    fieldConstructorClashing.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    FieldConstructorClashing readFieldConstructorClashing{reader};
    ASSERT_EQ(fieldConstructorClashing, readFieldConstructorClashing);
}

} // namespace field_constructor_clashing
} // namespace structure_types

File: test/language/structure_types/python/FieldConstructorClashingTest.py
import zserio

import StructureTypes

# this test is mainly for C++, so just check that it is ok
class FieldConstructorClashingTest(StructureTypes.TestCase):
    def testWriteRead(self):
        fieldConstructorClashing = self._createFieldConstructorClashing()

        writer = zserio.BitStreamWriter()
        fieldConstructorClashing.write(writer)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readFieldConstructorClashing = self.api.FieldConstructorClashing.from_reader(reader)
        self.assertEqual(fieldConstructorClashing, readFieldConstructorClashing)

    def _createFieldConstructorClashing(self):
        return self.api.FieldConstructorClashing(
            [self.api.CompoundRead(self.api.Field(self.FIELD1), self.api.Field(self.FIELD2))],
            [
                self.api.CompoundPackingRead(
                    self.api.Field(self.FIELD1), self.api.Field(self.FIELD2), self.api.Field(self.FIELD3)
                )
            ]
        )

    FIELD1 = 1
    FIELD2 = 9
    FIELD3 = 5
